From a10192e7c3237a707e9bb712a566e2f14415dddd Mon Sep 17 00:00:00 2001
From: Alan Coopersmith <alan.coopersmith@oracle.com>
Date: Sat, 10 Oct 2026 12:53:56 -0700
Subject: [PATCH] Revert "damage: Unlink a damage from the list it was actually
 inserted on"

This reverts commit 0fb16f3da03c077f5821f0eeb745384f5472faa5.

I missed that it broke driver ABI when I cherry-picked it from main to the
server-21.1-branch.   Apologies.

Closes: #1930
Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>
Part-of: <https://gitlab.freedesktop.org/xorg/xserver/-/merge_requests/2355>
---
 miext/damage/damage.c    | 22 +++++-----------------
 miext/damage/damagestr.h |  6 ------
 2 files changed, 5 insertions(+), 23 deletions(-)

diff --git a/miext/damage/damage.c b/miext/damage/damage.c
index f08aa59154..f3ae4ebbc9 100644
--- a/miext/damage/damage.c
+++ b/miext/damage/damage.c
@@ -1457,7 +1457,6 @@ damageRemoveDamage(DamagePtr * pPrev, DamagePtr pDamage)
     while (*pPrev) {
         if (*pPrev == pDamage) {
             *pPrev = pDamage->pNext;
-            pDamage->pListHead = NULL;
             return;
         }
         pPrev = &(*pPrev)->pNext;
@@ -1468,16 +1467,6 @@ damageRemoveDamage(DamagePtr * pPrev, DamagePtr pDamage)
 #endif
 }
 
-/*
- * Link pDamage onto the list headed by pPrev, remembering that list so that damageRemoveDamage() can later unlink it
- * from the same place.
- *
- * The list a window's damage belongs on is chosen by getDrawableDamageRef() and depends on the window pixmap, which can
- * change while the damage is registered (Composite redirection, rootless drawing and Present flips all swap it).
- * Re-deriving the list at removal time can therefore consult a different list than the damage was inserted onto, in
- * which case the removal silently does nothing and leaves an unregistered -- possibly freed -- damage linked where
- * damageRegionAppend() will dereference its NULL pDrawable.
- */
 static void
 damageInsertDamage(DamagePtr * pPrev, DamagePtr pDamage)
 {
@@ -1492,7 +1481,6 @@ damageInsertDamage(DamagePtr * pPrev, DamagePtr pDamage)
 #endif
     pDamage->pNext = *pPrev;
     *pPrev = pDamage;
-    pDamage->pListHead = pPrev;
 }
 
 static Bool
@@ -1565,9 +1553,11 @@ damageSetWindowPixmap(WindowPtr pWindow, PixmapPtr pPixmap)
     damageScrPriv(pScreen);
 
     if ((pDamage = damageGetWinPriv(pWindow))) {
+        PixmapPtr pOldPixmap = (*pScreen->GetWindowPixmap) (pWindow);
+        DamagePtr *pPrev = getPixmapDamageRef(pOldPixmap);
+
         while (pDamage) {
-            if (pDamage->pListHead)
-                damageRemoveDamage(pDamage->pListHead, pDamage);
+            damageRemoveDamage(pPrev, pDamage);
             pDamage = pDamage->pNextWin;
         }
     }
@@ -1731,7 +1721,6 @@ DamageCreate(DamageReportFunc damageReport,
         return 0;
     pDamage->pNext = 0;
     pDamage->pNextWin = 0;
-    pDamage->pListHead = NULL;
     RegionNull(&pDamage->damage);
     RegionNull(&pDamage->pendingDamage);
 
@@ -1834,8 +1823,7 @@ DamageUnregister(DamagePtr pDamage)
 #endif
     }
     pDamage->pDrawable = 0;
-    if (pDamage->pListHead)
-        damageRemoveDamage(pDamage->pListHead, pDamage);
+    damageRemoveDamage(getDrawableDamageRef(pDrawable), pDamage);
 }
 
 void
diff --git a/miext/damage/damagestr.h b/miext/damage/damagestr.h
index 19624192eb..0d07872e56 100644
--- a/miext/damage/damagestr.h
+++ b/miext/damage/damagestr.h
@@ -35,12 +35,6 @@
 typedef struct _damage {
     DamagePtr pNext;
     DamagePtr pNextWin;
-    /*
-     * Head of the pNext list this damage is currently linked on, or NULL if it is not linked on any list.  Recorded at
-     * insertion time so that removal always unlinks from the list the damage actually went onto; see
-     * damageInsertDamage().
-     */
-    DamagePtr *pListHead;
     RegionRec damage;
 
     DamageReportLevel damageLevel;
-- 
GitLab


